lib/pg_eventstore/queries.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+require_relative 'query_builders/events_filtering_query'
+
+module PgEventstore
+  class Queries
+    attr_reader :connection, :serializer, :deserializer
+    private :connection, :serializer, :deserializer
+
+    # @param connection [PgEventstore::Connection]
+    # @param serializer [PgEventstore::EventSerializer]
+    # @param deserializer [PgEventstore::PgresultDeserializer]
+    def initialize(connection, serializer, deserializer)
+      @connection = connection
+      @serializer = serializer
+      @deserializer = deserializer
+    end
+
+    # @return [void]
+    def transaction
+      connection.with do |conn|
+        # We are inside a transaction already - no need to start another one
+        if [PG::PQTRANS_ACTIVE, PG::PQTRANS_INTRANS].include?(conn.transaction_status)
+          next yield
+        end
+
+        conn.transaction do
+          conn.exec("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+          yield
+        end
+      end
+    rescue PG::TRSerializationFailure, PG::TRDeadlockDetected => e
+      retry if e.connection.transaction_status == PG::PQTRANS_IDLE
+      raise
+    end
+
+    # Finds a stream in the database by the given Stream object
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::Stream, nil] persisted stream
+    def find_stream(stream)
+      find_sql = <<~SQL
+        SELECT * FROM streams WHERE streams.context = $1 AND streams.stream_name = $2 AND streams.stream_id = $3
+          LIMIT 1
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(find_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym)) if pgresult.ntuples == 1
+    end
+
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::RawStream] persisted stream
+    def create_stream(stream)
+      create_sql = <<~SQL
+        INSERT INTO streams (context, stream_name, stream_id) VALUES ($1, $2, $3) RETURNING *
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(create_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym))
+    end
+
+    # @return [PgEventstore::Stream] persisted stream
+    def find_or_create_stream(stream)
+      find_stream(stream) || create_stream(stream)
+    end
+
+    # @see PgEventstore::Client#read for more info
+    # @param stream [PgEventstore::Stream]
+    # @param options [Hash]
+    # @return [Array<PgEventstore::Event>]
+    def stream_events(stream, options)
+      exec_params = events_filtering_builder(stream, options).to_exec_params
+      pgresult = connection.with do |conn|
+        conn.exec_params(*exec_params)
+      end
+      deserializer.deserialize_many(pgresult)
+    end
+
+    # @param stream [PgEventstore::Stream] persisted stream
+    # @param event [PgEventstore::Event]
+    # @return [PgEventstore::Event]
+    def insert(stream, event)
```

Comment:
Method is too complex I think. Are you using rubocop?

Reply:
Suggest the solution please